cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+
+        # translate exit status
+        if my_status == GRB.OPTIMAL and self._objective_value is None:
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status == GRB.OPTIMAL and self._objective_value is not None:
+            self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+        elif my_status == GRB.INFEASIBLE:
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        else:  # another?
+            raise NotImplementedError(
+                f"Translation of gurobi status {my_status} to CPMpy status not implemented")  # a new status type was introduced, please report on github
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return(self.cpm_status)
+
+        # translate solution values (of user vars only)
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                solver_val = self.solver_var(cpm_var).X
+                if cpm_var.is_bool():
+                    cpm_var._value = bool(solver_val)
+                else:
+                    cpm_var._value = int(solver_val)
+
+        return has_sol
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        from gurobipy import GRB
+
+        if is_num(cpm_var):
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return 1-self.solver_var(cpm_var._bv)
+
+        # create if it does not exit
+        if not cpm_var in self._varmap:
+            if isinstance(cpm_var, _BoolVarImpl):
+                revar = self.gbi_model.addVar(vtype=GRB.BINARY, name=cpm_var.name)
+            elif isinstance(cpm_var, _IntVarImpl):
+                revar = self.gbi_model.addVar(cpm_var.lb, cpm_var.ub, vtype=GRB.INTEGER, name=str(cpm_var))
+            else:
+                raise NotImplementedError("Not a known var {}".format(cpm_var))
+            self._varmap[cpm_var] = revar
+            # Update model to make new vars visible for constraints
+            self.gbi_model.update()
+
+        # return from cache
+        return self._varmap[cpm_var]
+
+    # if TEMPLATE does not support objective functions, you can delete minimize()/maximize()/_make_numexpr()
+    def minimize(self, expr):
+        """
+            Minimize the given objective function
+
+            `minimize()` can be called multiple times, only the last one is used
+
+            (technical side note: any constraints created during conversion of the objective
+            are premanently posted to the solver)
+        """
+        from gurobipy import GRB
+
+        # make objective function non-nested
+        (flat_obj, flat_cons) = (flatten_objective(expr))
+        self += linearize_constraint(flat_cons)  # add potentially created constraints
```

Comment:
This linearizie not needed, this is done in __add__ anyway.

Given how for all solvers the minimize/maximize is so similar, should we perhaps make them superclass functions, and have one objective(self, expr, minimize=True) function instead to implement in the interfaces?

Reply:
That would make sense indeed. Difference between minimize and maximize is at most one line in all solvers so...